File: wiki/versions/1.19/entity/entity_attributes/src/main/java/org/quiltmc/wiki/entity_attributes/mixin/LivingEntityMixin.java
package org.quiltmc.wiki.entity_attributes.mixin;

import org.quiltmc.wiki.entity_attributes.AttributesExample;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.attribute.DefaultAttributeContainer;
import net.minecraft.entity.attribute.EntityAttribute;
import net.minecraft.entity.attribute.EntityAttributeInstance;
import org.jetbrains.annotations.Nullable;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Shadow;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

@Mixin(LivingEntity.class)
public abstract class LivingEntityMixin {

	// @start Add-Attribute-Mixin
	@Inject(method = "createLivingAttributes", at = @At(value = "RETURN"))
	private static void addCustomAttribute(CallbackInfoReturnable<DefaultAttributeContainer.Builder> cir) {
		cir.getReturnValue().add(AttributesExample.GENERIC_JUMP_BOOST);
	}
	// @end Add-Attribute-Mixin

	// @start Use-Attribute
	@Shadow @Nullable public abstract EntityAttributeInstance getAttributeInstance(EntityAttribute attribute);

	@Inject(method = "getJumpVelocity", at = @At(value = "RETURN"), cancellable = true)
	private void applyJumpBoost(CallbackInfoReturnable<Float> cir) {
		EntityAttributeInstance instance = getAttributeInstance(AttributesExample.GENERIC_JUMP_BOOST);
		cir.setReturnValue(cir.getReturnValue() * (float) instance.getValue());
	}
	// @end Use-Attribute
}

File: wiki/versions/1.19/entity/entity_attributes/scala/src/main/scala/org/quiltmc/wiki/entity_attributes/mixin/LivingEntityMixin.scala
package org.quiltmc.wiki.entity_attributes.mixin

import net.minecraft.entity.LivingEntity
import net.minecraft.entity.attribute.{DefaultAttributeContainer, EntityAttribute, EntityAttributeInstance}
import org.quiltmc.wiki.entity_attributes.AttributesExample
import org.spongepowered.asm.mixin.injection.{At, Inject}
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable
import org.spongepowered.asm.mixin.{Mixin, Shadow}

import scala.annotation.static

@Mixin(Array(classOf[LivingEntity]))
abstract class LivingEntityMixin {

    // @start Use-Attribute
    @Shadow
    def getAttributeInstance(attribute: EntityAttribute): EntityAttributeInstance

    @Inject(method = Array("getJumpVelocity"), at = Array(At(value = "RETURN")), cancellable = true)
    private def applyJumpBoost(cir: CallbackInfoReturnable[Float]): Unit = Option(
        getAttributeInstance(AttributesExample.GENERIC_JUMP_BOOST)
    ).foreach { instance =>
        cir.setReturnValue(cir.getReturnValue * instance.getValue.toFloat)
    }
    // @end Use-Attribute
}
object LivingEntityMixin {

    // @start Add-Attribute-Mixin
    @static
    @Inject(method = Array("createLivingAttributes"), at = Array(At(value = "RETURN")))
    private def addCustomAttribute(cir: CallbackInfoReturnable[DefaultAttributeContainer.Builder]): Unit =
        cir.getReturnValue.add(AttributesExample.GENERIC_JUMP_BOOST)
    // @end Add-Attribute-Mixin
}
